EnableModExpansion: allows overriding scale/speed inside ESL entries

emoose opened the pull request:
Separated the ESL expansion stuff from #266, since I'm probably gonna redo that PR eventually, and I guess it's better to get this stuff in sooner than later.

This adds a `EnableModExpansion` setting which enables the hooks for allowing scale/speed to be set inside ESL, had to change how the flag for enabling speed/scale is set though, since it seems 0xA might actually be in use (I didn't see any code using it, but the field is defined)
Now instead of setting 0xA to 1, you'll need to set the first nibble of it to F instead.
eg. if 0xA was set to `03`, you'd set it to `F3`, or if 0xA is `00` just set it to `F0`.
With that set the last 4 bytes of the ESL entry will be treated as two uint16_t percentage values, first one as speed percent, second as scale percent.

Also added stuff to allow scaling in ETS as well, as requested at #303, more info in the comment below, seems to be working fine from what I've heard 😸 

Still need to fix the interaction issue mentioned below though... hopefully can figure something out for it soon.

FutonGama commented:
I tested the Speed and Size and its working nice like the old build. I need to test now the ETS size.

emoose commented:
Added code to also scale the ETS models collision radius/height as well, many thanks to @Corazon-SA for helping to test & refine it 👍 

We noticed that some objects have weird collision where the radius begins at an offsetted position inside the models, seems doors use this a lot - fortunately scaling that offset by the collision scale value _mostly_ seems to get collision to match up with the scale, but sometimes this scaled-offset isn't that perfect (maybe the original values weren't an exact match in the beginning, and now scaling them up makes that more apparent..)

To get around that I added a flag which lets you specify the model scale & collision scale separately, so modders can fine-tune it if they need to. 
The ETS_DATA struct was pretty much all occupied already, but it seems the `EtcModelNo_2` field (uint16_t) is actually restricted by the game to values 0x40 and below, so with a single patch we can make the game treat it as a byte instead, giving us an extra byte to use as a flag.

With the flag set the unused 12-byte scale area gets treated as two 6-byte vectors, one for model scale & one for collision scale, the vectors are 2-byte shorts for X Y Z, values are specified as percentages (so for 2x scale you'd set it as 200 / 200 / 200)
The flag is at 0x3 into the ETS struct, setting that to `F0` will enable this "short mode".

Some tools already treat the 12-byte scale area as a normal 12-byte vector, those can still be used with this fine, as long as that flag isn't set then both model & collision will get the values taken from there.

Getting late here and I'm pretty tired, I'll update OP with a better description tomorrow :)

FutonGama commented:
Its working, and the Collision Build too, ETS with collision working perfectly. Thank you. Darth is without computer so i'm testing it alone.

emoose commented:
Well looking into the checks for window "jump out" prompt to appear it seems `cPlayer::actionSelect` is responsible for the prompt itself, which calls a bunch of check funcs to see if you're near a window/door/etc.

`windowCheck` calls into `ChkWindow`, which seems to use `SceAtCheckFieldInfo` to find an AEV related to the window, that finally calls into `sceAtHitCheck` which seems to be what actually checks if you're near it, looks like AEV has `hit_dir_ang_48` / `hit_open_ang_49` fields that it uses to check against, I guess maybe modders can just update those ang fields if they need to, doesn't seem like there's much else we can change there at least...

Should be fine to merge, I'll add the debug build fix here in a sec, and maybe add some more info about what mod expansion does into settings_string.h too.

>Why is this, exactly? Anything broken there?
Hmm not really sure, haven't tried building it with the latest changes in a while, there are some commits in there like the column helper funcs that I'd guess might conflict though, not really sure if I can just remove those from the branch or not...

I've been thinking about redoing the Em manager side of it for a while though, would be nice if that could be 'popped-out' as a separate window so that you could see it while playing, kinda like the debug stuff added to ocarina of time here: https://youtu.be/r1BPWSADWck?t=206
(I like the way they added a toolbar menu at the top there too, if we ever added more debug-window kinda stuff it might be nice to add something like that too)

emoose commented:
>Should be fine to merge, I'll add the debug build fix here in a sec, and maybe add some more info about what mod expansion does into settings_string.h too.

Pushed, also checked with the usual EXEs, seems to be fine.

nipkownix commented:
> Hmm not really sure, haven't tried building it with the latest changes in a while, there are some commits in there like the column helper funcs that I'd guess might conflict though, not really sure if I can just remove those from the branch or not...

Ah, I rebased the branch some time ago, so I think there should be only minor conflicts? Haven't checked again since then.

> but it'd be nice if that could be 'popped-out' as a separate window so that you could see it while playing, kinda like the debug stuff added to ocarina of time here: https://youtu.be/r1BPWSADWck?t=203

Oh, that seems nice. We can definitely add an option to pop out the window. Not sure how we'd handle passing input to it, though. Unless you think the game shouldn't have input passed to it as long as the pop up is open?

> Pushed, also checked with the usual EXEs, seems to be fine.

Many thanks!

emoose commented:
> Unless you think the game shouldn't have input passed to it as long as the pop up is open?

I was thinking of it more like an overlay kinda, eg. you'd first open it from cfgMenu, be able to press buttons/select Ems/etc while still in the cfgMenu "mode" to choose what you want to monitor, but closing down cfgMenu would still leave the Em/debug window on screen, with inputs going to the game as normal.
Then if you need to adjust anything on that window (or move it/close it down) you can just bring up cfgMenu again, which would let inputs go to all the windows again.

Not sure how possible that is with the current setup, we do already have the console window drawing without cfgMenu needing to be shown at least, so I guess displaying it shouldn't be too hard, not sure about inputs though.

FutonGama commented:
Well if you guys need help to test something i'm here to do the necessary, every new features feels very good to me, loved. Count on me.
